google-cloud-datastore/lib/google/cloud/datastore/dataset/query_results.rb

```diff
@@ -375,27 +388,39 @@ def all_with_cursor request_limit: nil, &block
             end
           end
 
+          # rubocop:disable Metrics/AbcSize
+
           ##
           # @private New Dataset::QueryResults from a
           # Google::Dataset::V1::RunQueryResponse object.
-          def self.from_grpc query_res, service, namespace, query, read_time = nil
-            r, c = Array(query_res.batch.entity_results).map do |result|
-              [Entity.from_grpc(result.entity), Cursor.from_grpc(result.cursor)]
-            end.transpose
+          def self.from_grpc query_res, service, namespace, query, read_time = nil, explain_options = nil
+            if query_res.batch
+              r, c = Array(query_res.batch.entity_results).map do |result|
+                [Entity.from_grpc(result.entity), Cursor.from_grpc(result.cursor)]
+              end.transpose
+            end
             r ||= []
             c ||= []
+
+            next_more_results = query_res.batch.more_results if query_res.batch
+            next_more_results ||= :NO_MORE_RESULTS
+
             new(r).tap do |qr|
               qr.cursors = c
-              qr.end_cursor = Cursor.from_grpc query_res.batch.end_cursor
-              qr.more_results = query_res.batch.more_results
+              qr.explain_metrics = query_res.explain_metrics
+              qr.end_cursor = Cursor.from_grpc query_res.batch.end_cursor if query_res.batch
+              qr.more_results = next_more_results
               qr.service = service
               qr.namespace = namespace
               qr.query = query_res.query || query
               qr.instance_variable_set :@read_time, read_time
-              qr.instance_variable_set :@batch_read_time, query_res.batch.read_time
+              qr.instance_variable_set :@batch_read_time, query_res.batch.read_time if query_res.batch
+              qr.instance_variable_set :@explain_options, explain_options
```

Comment:
Just use a private `attr_writer` instead of this `instance_variable_set`. The latter is brittle and not very ruby-like. I know we have existing code that does this, but I hate it and want to clean it up as well at some point.